Model/Product/View/ProductViewFactory.php

```diff
@@ -39,37 +48,66 @@ class ProductViewFactory implements ProductViewFactoryInterface
+        $this->productInformationRepository = $productInformationRepository;
         $this->routableResourceRepository = $routableResourceRepository;
         $this->contentViewFactory = $contentViewFactory;
     }
 
-    public function create(ProductInformationInterface $product, array $dimensions): ProductViewInterface
+    /**
+     * @param DimensionInterface[] $dimensions
+     */
+    public function create(ProductInterface $product, array $dimensions): ProductViewInterface
     {
-        $viewProduct = new ProductView($product->getProductId(), $product->getProductCode());
-        $viewProduct->setProductInformation($product);
+        $productInformation = null;
+        $routableResource = null;
+        $locale = null;
 
-        $contentDimensions = $this->contentRepository->findByDimensions(
-            ProductInterface::RESOURCE_KEY,
-            $product->getProductId(),
-            $dimensions
-        );
+        foreach ($dimensions as $dimension) {
+            if (null === $productInformation) {
+                $productInformation = $this->productInformationRepository->findByProductId(
+                    $product->getId(),
+                    $dimension
+                );
+            }
 
-        $content = $this->contentViewFactory->create($contentDimensions);
-        if ($content) {
-            $viewProduct->setContent($content);
+            if (null === $routableResource) {
+                $routableResource = $this->routableResourceRepository->findByResource(
+                    ProductInterface::RESOURCE_KEY,
+                    $product->getId(),
+                    $dimension
+                );
+            }
+
+            if (null === $locale && $dimension->hasAttribute(DimensionInterface::ATTRIBUTE_KEY_LOCALE)) {
+                $locale = $dimension->getAttributeValue(DimensionInterface::ATTRIBUTE_KEY_LOCALE);
+            }
+        }
+        if (null === $productInformation) {
+            throw new ProductInformationNotFoundException($product->getCode());
+        }
+        if (null === $routableResource) {
+            throw new RoutableResourceNotFoundException(ProductInterface::RESOURCE_KEY, $product->getId());
+        }
+        if (null === $locale) {
+            throw new \InvalidArgumentException('No locale found');
         }
 
-        $routableResource = $this->routableResourceRepository->findOrCreateByResource(
+        $content = $this->contentViewFactory->loadAndCreate(
             ProductInterface::RESOURCE_KEY,
-            $product->getProductId(),
-            $product->getDimension()
+            $product->getId(),
+            $dimensions
         );
-        $viewProduct->setRoutableResource($routableResource);
+        if (null === $content) {
+            throw new ContentNotFoundException(ProductInterface::RESOURCE_KEY, $product->getId());
+        }
+
+        $viewProduct = new ProductView($locale, $product, $productInformation, $content, $routableResource);
```

Comment:
👍 this looks good now - what do you think?

Reply:
At least it's working :+1: